### PR TITLE
Add keycloak integration test

### PR DESCRIPTION
**Pull request by corneil**
Adds Keycloak integration test to DataflowOAuthIT
Adds Authorities mapping test similar to keycloak role usage.
Added scripts to src/local for testing keycloak locally with preconfigured roles / group and user.

## NOTES
The Keycloak configuration uses roles and assigns user to a group containing those roles.
The configuration supports both basic auth endpoint like needed by dataflow shell and REST calls and OAuth2 flow needed by the UI. Default user in keycloak realm is `joe:password`.

### CODE REVIEW

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/db/container/DataflowCluster.java [hunk 1]
```diff
@@ -396,32 +414,61 @@ private SkipperContainer<?> buildSkipperContainer(
...
-			skipperContainer.withEnv("SPRING_SECURITY_OAUTH2_CLIENT_PROVIDER_UAA_JWK_SET_URI", "http://oauth:8099/uaa/token_keys");
-			skipperContainer.withEnv("SPRING_SECURITY_OAUTH2_CLIENT_PROVIDER_UAA_TOKEN_URI", "http://oauth:8099/uaa/oauth/token");
-			skipperContainer.withEnv("SPRING_SECURITY_OAUTH2_CLIENT_PROVIDER_UAA_USER_INFO_URI", "http://oauth:8099/uaa/userinfo");
-			skipperContainer.withEnv("SPRING_SECURITY_OAUTH2_CLIENT_PROVIDER_UAA_USER_NAME_ATTRIBUTE", "user_name");
-			skipperContainer.withEnv("SPRING_SECURITY_OAUTH2_CLIENT_PROVIDER_UAA_AUTHORIZATION_URI", "http://oauth:8099/uaa/oauth/authorize");
-			skipperContainer.withEnv("SPRING_SECURITY_OAUTH2_RESOURCESERVER_OPAQUETOKEN_INTROSPECTION_URI", "http://oauth:8099/uaa/introspect");
-			skipperContainer.withEnv("SPRING_SECURITY_OAUTH2_RESOURCESERVER_OPAQUETOKEN_CLIENT_ID", "dataflow");
-			skipperContainer.withEnv("SPRING_SECURITY_OAUTH2_RESOURCESERVER_OPAQUETOKEN_CLIENT_SECRET", "secret");
-			skipperContainer.withEnv("SPRING_SECURITY_OAUTH2_AUTHORIZATION_CHECK_TOKEN_ACCESS", "isAuthenticated()");
+			if(oauthContainer.getDockerImageName().contains(TagNames.UAA)) {
```
Comment (onobc) [inline spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/db/container/DataflowCluster.java:417]: It is really hard to tell if anything changed here in the old path and a super large if/else is not fun to read through. Wdyt about factoring this out a bit to something like:
```java
if (id.startsWith(TagNames.KEYCLOAK)) {
   configureKeycloak(skipperContainer);
}
else {
   configureUaa(skipperContainer);
}
```

> [!NOTE]
> For consistency, since we already use `id.startsWith(TagNames.KEYCLOAK))` to differentiate the 2, why not just use it here rather than dig around in the docker container?

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/db/container/DataflowCluster.java [hunk 2]
```diff
@@ -396,32 +414,61 @@ private SkipperContainer<?> buildSkipperContainer(
 		}
 
 		if (oauthContainer != null) {
```
Comment (onobc) [inline spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/db/container/DataflowCluster.java:416]: Why were we checking if `oauthContainer != null` anyways? We should assert it is not null way above otherwise there is something else seriously wrong w/ this test.

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/oauth/DataflowOAuthIT.java [hunk 3]
```diff
@@ -92,4 +92,60 @@ void securedSetup() throws Exception {
...
+					.ignoreExceptions()
+					.atMost(90, TimeUnit.SECONDS)
+					.untilAsserted(() -> {
+						log.info("Checking auth using curl");
+						ExecResult cmdResult = execInToolsContainer("curl", "-v", "-u", "joe:password", "http://dataflow:9393/about");
+						String response = cmdResult.getStdout();
+						if (StringUtils.hasText(response)) {
+							log.info("Response is {}", response);
+						}
+						if(StringUtils.hasText(cmdResult.getStderr())) {
```
Comment (onobc) [inline spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/oauth/DataflowOAuthIT.java:122]: Again, inconsistent spacing line 122 and 119. please adjust in all newly added code

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/oauth/DataflowOAuthIT.java [hunk 4]
```diff
@@ -92,4 +92,60 @@ void securedSetup() throws Exception {
 			}
 		}
 	}
+
+	@Test
+	void securedSetupKeycloak() throws Exception {
+		log.info("Running securedSetupKeycloak()");
+		this.dataflowCluster.startIdentityProvider(TagNames.KEYCLOAK_26);
```
Comment (onobc) [inline spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/oauth/DataflowOAuthIT.java:99]: What is the significance of KEYCLOAK_26?

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/oauth/DataflowOAuthIT.java [hunk 5]
```diff
@@ -92,4 +92,60 @@ void securedSetup() throws Exception {
...
+			if (StringUtils.hasText(response)) {
+				log.info("Response is {}", response);
+			}
+			response = cmdResult.getStderr();
+			if(StringUtils.hasText(response)) {
+				log.warn("Error is {}", response);
+			}
+			stderr.set(cmdResult.getStderr());
+			assertThat(cmdResult.getExitCode()).isNotZero();
+			stderr.set("");
```
Comment (onobc) [inline spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/oauth/DataflowOAuthIT.java:142]: Can we factor this chunk of code out for DRY?

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/oauth/DataflowOAuthIT.java [hunk 6]
```diff
@@ -92,4 +92,60 @@ void securedSetup() throws Exception {
...
+		}
+		finally {
```
Comment (onobc) [inline spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/oauth/DataflowOAuthIT.java:144]: Why is this needed here?

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/oauth/DataflowOAuthIT.java [hunk 7]
```diff
@@ -92,4 +92,60 @@ void securedSetup() throws Exception {
 			}
 		}
 	}
+
+	@Test
+	void securedSetupKeycloak() throws Exception {
```
Comment (onobc) [inline spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/oauth/DataflowOAuthIT.java:97]: Why is this called "..Setup..." ? Can you add a comment as to what this is testing?

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/unit/test/DataFlowAuthenticationTests.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+import org.springframework.cloud.dataflow.server.single.DataFlowServerApplication;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.test.context.ActiveProfiles;
+import org.springframework.test.context.DynamicPropertyRegistry;
+import org.springframework.test.context.DynamicPropertySource;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@ActiveProfiles("keycloak")
+@SpringBootTest(classes = { DataFlowServerApplication.class },
```
Comment (onobc) [inline spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/unit/test/DataFlowAuthenticationTests.java:31]: Since only 1 class we can use compact form w/o `{ }`.

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/unit/test/DataFlowAuthenticationTests.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+		webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+@Testcontainers
+@Disabled("Determine how to run app and test client in different contexts")
+public class DataFlowAuthenticationTests {
+
+	private static final Logger logger = LoggerFactory.getLogger(DataFlowAuthenticationTests.class);
+
+	@Container
+	static KeycloakContainer keycloakContainer = new KeycloakContainer("keycloak/keycloak:25.0")
```
Comment (onobc) [inline spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/unit/test/DataFlowAuthenticationTests.java:40]: Why not use the TAGNAMES constants?

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/unit/test/DataFlowAuthenticationTests.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+			}
+		});
+
+	@DynamicPropertySource
+	static void configureProperties(DynamicPropertyRegistry registry) {
+		registry.add("keycloak.url", keycloakContainer::getAuthServerUrl);
+	}
+
+	@Test
+	void testAuthentication() throws Exception {
```
Comment (onobc) [inline spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/unit/test/DataFlowAuthenticationTests.java:61]: Do you mind giving this method name a bit more detail?

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/unit/test/DataFlowAuthenticationTests.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+		finally {
+			CommandLineApp.completed.set(true);
+		}
+	}
+
+	@SpringBootApplication
+	@ImportAutoConfiguration(DataFlowClientAutoConfiguration.class)
+	public static class CommandLineApp implements CommandLineRunner {
+
+		public static AtomicBoolean completed = new AtomicBoolean(false);
```
Comment (onobc) [inline spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/unit/test/DataFlowAuthenticationTests.java:87]: Using a `CountdownLatch` would allow to do a `latch.await` rather than an indefinite awaitility await.

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/unit/test/DataFlowAuthenticationTests.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+	@DynamicPropertySource
+	static void configureProperties(DynamicPropertyRegistry registry) {
+		registry.add("keycloak.url", keycloakContainer::getAuthServerUrl);
+	}
+
+	@Test
+	void testAuthentication() throws Exception {
+		try (ConfigurableApplicationContext applicationContext = SpringApplication.run(CommandLineApp.class,
+				"--spring.profiles.active=keycloak-client",
```
Comment (onobc) [inline spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/unit/test/DataFlowAuthenticationTests.java:63]: I guess this is related to the comment in the `@Disabled` as you are trying to use `@SBT` to start the server and then use this manual auth context to start the client, correct?

File: src/carvel/config/values/values.yml [hunk 13]
```diff
@@ -112,5 +112,5 @@ scdf:
         enabled: false
         image:
           repository: micrometermetrics/prometheus-rsocket-proxy
-          tag: 2.0.0-M2
+          tag: 1.5.3
```
Comment (onobc) [inline src/carvel/config/values/values.yml:115]: Why is this in here? We are not downgrading PRP and if we were we would not do it in this PR to add an IT.

File: src/local/launch-dataflow.sh [hunk 14]
```diff
@@ -16,7 +16,7 @@ SCDIR=$(dirname "$(readlink -f "${BASH_SOURCE[0]}")")
 SCDIR=$(realpath $SCDIR)
 PROJECT_DIR=$(realpath "$SCDIR/../..")
 if [ "$DATAFLOW_VERSION" = "" ]; then
-    DATAFLOW_VERSION=2.11.3-SNAPSHOT
+    DATAFLOW_VERSION=3.0.0-SNAPSHOT
```
Comment (onobc) [inline src/local/launch-dataflow.sh:19]: Is this needed for the Keycloak IT?

**Review by onobc — CHANGES_REQUESTED** [review]
Thanks for adding this @corneil . I have quite a few comments/suggestions but all in all this is looking good.

File: .github/workflows/ci-it-security.yml [hunk 15]
```diff
@@ -33,7 +33,6 @@ jobs:
     - name: Run Security IT
       shell: bash
       run: |
-        ./mvnw clean install -DskipTests -T 1C -s .settings.xml -pl spring-cloud-dataflow-server -am  -B --no-transfer-progress
```
[the remark this answers is not in the document]
Comment (corneil) [reply .github/workflows/ci-it-security.yml:36]: This builds what is needed to execute the test.

File: spring-cloud-common-security-config/spring-cloud-common-security-config-web/src/test/java/org/springframework/cloud/common/security/support/DefaultAuthoritiesMapperTests.java [hunk 16]
```diff
@@ -93,6 +93,33 @@ public void testMapConstructorWithIncompleteRoleMappings() throws Exception {
 	}
 
 	@Test
+	public void testThat3MappedAuthoritiesAreReturned() throws Exception {
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-common-security-config/spring-cloud-common-security-config-web/src/test/java/org/springframework/cloud/common/security/support/DefaultAuthoritiesMapperTests.java:96]: Done

File: spring-cloud-dataflow-server/src/main/resources/application.yml [hunk 17]
```diff
@@ -28,6 +28,7 @@ spring:
             url: "{repository}/org/springframework/cloud/spring-cloud-dataflow-shell/{version}/spring-cloud-dataflow-shell-{version}.jar"
             checksum-sha1-url: "{repository}/org/springframework/cloud/spring-cloud-dataflow-shell/{version}/spring-cloud-dataflow-shell-{version}.jar.sha1"
             checksum-sha256-url: "{repository}/org/springframework/cloud/spring-cloud-dataflow-shell/{version}/spring-cloud-dataflow-shell-{version}.jar.sha256"
+
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-server/src/main/resources/application.yml:31]: I'll have to rewrite/force push the whole change to revert this.

File: spring-cloud-common-security-config/spring-cloud-common-security-config-web/src/test/java/org/springframework/cloud/common/security/support/DefaultAuthoritiesMapperTests.java [hunk 18]
```diff
@@ -93,6 +93,33 @@ public void testMapConstructorWithIncompleteRoleMappings() throws Exception {
...
+		Map<String, String> roleMappings = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-common-security-config/spring-cloud-common-security-config-web/src/test/java/org/springframework/cloud/common/security/support/DefaultAuthoritiesMapperTests.java:97]: Refactoring

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/db/container/DataflowCluster.java [hunk 19]
```diff
@@ -190,11 +193,26 @@ public void startIdentityProvider(String id) {
 		ClusterContainer clusterContainer = this.oauthImages.get(id);
 		Assert.notNull(clusterContainer, String.format("Unknown oauth %s", id));
 
-		GenericContainer<?> oauthContainer = new GenericContainer<>(clusterContainer.image);
+		GenericContainer<?> oauthContainer;
+		if(id.startsWith(TagNames.KEYCLOAK)) {
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/db/container/DataflowCluster.java:197]: Done

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/oauth/DataflowOAuthIT.java [hunk 20]
```diff
@@ -40,8 +40,8 @@ class DataflowOAuthIT extends AbstractDataflowTests {
 	private final Logger log = LoggerFactory.getLogger(DataflowOAuthIT.class);
 
 	@Test
-	void securedSetup() throws Exception {
-		log.info("Running testSecuredSetup()");
+	void securedSetupUAA() throws Exception {
+		log.info("Running securedSetupUAA()");
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/oauth/DataflowOAuthIT.java:44]: good suggestion

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/db/container/DataflowCluster.java (hunk 2 above)
Comment (corneil) [reply spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/db/container/DataflowCluster.java:416]: The DataflowCluster is used to execute database tests where there are no oauthContainer

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/oauth/DataflowOAuthIT.java (hunk 4 above)
Comment (corneil) [reply spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/oauth/DataflowOAuthIT.java:99]: KEYCLOAK 26 is latest release. I tested with bothh 25.0 and 26.0 to make sure we cover the last 2 versions.

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/oauth/DataflowOAuthIT.java (hunk 6 above)
Comment (corneil) [reply spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/oauth/DataflowOAuthIT.java:144]: We only dump all the error logs when there is an error.

File: src/local/launch-dataflow.sh (hunk 14 above)
Comment (corneil) [reply src/local/launch-dataflow.sh:19]: for any local testing

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/db/AbstractDataflowTests.java [hunk 21]
```diff
@@ -91,8 +90,10 @@ protected static class EmptyConfig {
 		ClusterContainer.from(TagNames.DB2_11_5_8_0, "icr.io/db2_community/db2:11.5.8.0", TagNames.DB2)
 	);
 
-	public final static List<ClusterContainer> OAUTH_CONTAINERS = Collections.singletonList(
-		ClusterContainer.from(TagNames.UAA_4_32, "springcloud/scdf-uaa-test:4.32", TagNames.UAA)
+	public final static List<ClusterContainer> OAUTH_CONTAINERS = Arrays.asList(
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/db/AbstractDataflowTests.java]: Refactored

File: spring-cloud-common-security-config/spring-cloud-common-security-config-web/src/test/java/org/springframework/cloud/common/security/support/DefaultAuthoritiesMapperTests.java [hunk 22]
```diff
@@ -93,6 +93,33 @@ public void testMapConstructorWithIncompleteRoleMappings() throws Exception {
...
+		roles.add("dataflow_view");
+		roles.add("dataflow_deploy");
+
+		DefaultAuthoritiesMapper defaultAuthoritiesMapper = new DefaultAuthoritiesMapper("uaa", providerRoleMapping);
+		Collection<? extends GrantedAuthority> authorities = defaultAuthoritiesMapper.mapScopesToAuthorities("uaa",
+				roles, null);
+
+		assertThat(authorities).hasSize(3);
+		assertThat(authorities.stream().map(authority -> authority.getAuthority()).collect(Collectors.toList()))
+				.containsExactlyInAnyOrder("ROLE_DEPLOY", "ROLE_MANAGE", "ROLE_VIEW");
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-common-security-config/spring-cloud-common-security-config-web/src/test/java/org/springframework/cloud/common/security/support/DefaultAuthoritiesMapperTests.java]: Refactored other instances as well.

File: src/local/README.md [hunk 23]
```diff
@@ -20,3 +20,23 @@ This script requires [jib-cli](https://github.com/GoogleContainerTools/jib/tree/
...
+```
+### Execute:
+```shell
+./src/local/launch-keycloak.sh
+```
+
+### Debug
+Module: `spring-cloud-dataflow-server`
+Class: `org.springframework.cloud.dataflow.server.single.DataFlowServerApplication`
+Arguments: `--spring.cloud.dataflow.features.streams-enabled=false --spring.cloud.dataflow.features.tasks-enabled=true --spring.cloud.dataflow.features.schedules-enabled=false --spring.config.additional-location="$ProjectFileDir$/src/local/application-dataflow-keycloak.yaml"`
```
Comment (cppwfs) [inline src/local/README.md:38]: Just curious as to why we are disabling streams?

File: src/local/application-dataflow-keycloak.yaml [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+            authorization-grant-type: 'authorization_code'
+#            client-authentication-method: # unsure of value
+            scope:
+              - openid
+              - roles
+      resourceserver:
+        opaquetoken:
+          introspection-uri: http://localhost:8080/realms/dataflow/protocol/openid-connect/token/introspect
+          client-id: 'dataflow'
+          client-secret: '090RucamvekrMLyGHMr4lkHX9xhAlsqK'
```
Comment (cppwfs) [inline src/local/application-dataflow-keycloak.yaml:45]: We need a note here to tell folks to use their own secret here.

File: src/local/launch-keycloak.sh [hunk 25]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+SCDIR=$(dirname "$(readlink -f "${BASH_SOURCE[0]}")")
+SCDIR=$(realpath $SCDIR)
+
+docker run --name keycloak  -p 8080:8080 -p 9000:9000 \
```
Comment (cppwfs) [inline src/local/launch-keycloak.sh:5]: Curious.   Why does the docker compose version require a database and this one does not?

File: src/local/launch-with-keycloak.sh [hunk 26]
```diff
@@ -0,0 +1,93 @@
...
+  --spring.datasource.password=spring \
+  --spring.datasource.driver-class-name=org.mariadb.jdbc.Driver \
+  --spring.jpa.database.platform=org.hibernate.dialect.MariaDB106Dialect \
+  --spring.config.additional-location="file://$SCDIR/application-dataflow-keycloak.yaml" $SCDF_ARGS | tee server.log &
+
+echo "$!" > dataflow.pid
+echo "Waiting for Data Flow Server"
+wget --retry-connrefused --read-timeout=20 --timeout=15 --tries=10 --continue -q http://localhost:9393
+echo "Data flow running"
+# docker run -v ${SCDIR}/data:/opt/keycloak/data/import -p 8080:8080 -e KC_BOOTSTRAP_ADMIN_USERNAME=admin -e KC_BOOTSTRAP_ADMIN_PASSWORD=admin keycloak/keycloak:25.0 start-dev --import-realm --verbose
```
Comment (cppwfs) [inline src/local/launch-with-keycloak.sh:92]: We can remove commented line.

**Review by cppwfs — CHANGES_REQUESTED** [review]

File: spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/unit/test/DataFlowAuthenticationTests.java (hunk 8 above)
Comment (corneil) [reply spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/unit/test/DataFlowAuthenticationTests.java:31]: Deleting in favour of proper integration test.

File: src/carvel/config/values/values.yml (hunk 13 above)
Comment (corneil) [reply src/carvel/config/values/values.yml]: I can revert these with a force push once all other changes are ready,

File: src/local/README.md (hunk 23 above)
Comment (corneil) [reply src/local/README.md:38]: This was for testing login when only running Keycloak and Data Flow Server.

File: src/local/application-dataflow-keycloak.yaml (hunk 24 above)
Comment (corneil) [reply src/local/application-dataflow-keycloak.yaml:44]: That secret is specific to the realm we import when using the container.

File: src/local/launch-with-keycloak.sh (hunk 26 above)
Comment (corneil) [reply src/local/launch-with-keycloak.sh]: Done

**Review by onobc — APPROVED** [review]
Thanks for the updates @corneil - LGTM.

**Comment by corneil** [comment]
Squashed all changes into one commit which allowed dropping some unneeded changes.

File: src/local/launch-keycloak.sh (hunk 25 above)
Comment (corneil) [reply src/local/launch-keycloak.sh:5]: Keycloak will use an embedded in memory database if you don't provide jdbc credentials.

Comment (cppwfs) [reply src/local/launch-keycloak.sh:5]: Thank you!!!!
